src/charm.py

```diff
@@ -134,55 +69,35 @@ def _reconcile(self):
             initialization and the event emission.
         """
         container = self.unit.get_container(self._container_name)
-        charm_root = self.charm_dir.absolute()
-        forward_alert_rules = cast(bool, self.config.get("forward_alert_rules"))
         insecure_skip_verify = cast(bool, self.config.get("tls_insecure_skip_verify"))
 
         # Integrate with TLS relations
-        receive_ca_certs_hash = self._integrate_receive_ca_cert(container)
-
-        server_cert_hash = self._integrate_server_cert(container)
-
-        self.config_manager = ConfigManager(
-            receiver_tls=is_server_cert_on_disk(container),
-            insecure_skip_verify=cast(bool, self.config.get("tls_insecure_skip_verify")),
-        )
-
-        # Dashboards setup
-        self._forward_dashboards()
-
-        # Logs setup
-        loki_provider = LokiPushApiProvider(
+        receive_ca_certs_hash = integrations.receive_ca_cert(
             self,
-            relation_name="receive-loki-logs",
-            port=Port.loki_http.value,
-            scheme="https" if is_server_cert_on_disk(container) else "http",
+            recv_ca_cert_folder_path=ContainerPath(RECV_CA_CERT_FOLDER_PATH, container=container),
+            refresh_certs=partial(refresh_certs, container=container),
         )
-        loki_consumer = LokiPushApiConsumer(
+        server_cert_hash = integrations.receive_server_cert(
```

Comment:
Going forward we should think about a neat way to formalize the return types of these `integrate_x` methods.